File: 백준/Gold/16946. 벽 부수고 이동하기 4/벽 부수고 이동하기 4.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static int N;
    static int M;
    static int[][] map;
    static boolean[][] visited;

    static int[] dr = new int[]{-1, 0, 1, 0};
    static int[] dc = new int[]{0, -1, 0, 1};


    static Node[][] result;

    public static class Node {
        int cnt;
        int num;

        public Node(int cnt, int num) {
            this.cnt = cnt;
            this.num = num;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        map = new int[N][M];
        visited = new boolean[N][M];
        result = new Node[N][M];


        for (int n = 0; n < N; n++) {
            String str = br.readLine();
            for (int m = 0; m < M; m++) {
                map[n][m] = str.charAt(m) - '0';
                result[n][m] = new Node(0,0);

            }
        }

        int num = 2;
        for (int i = 0; i < N; i++) {
            for(int j=0; j<M; j++){
                if(map[i][j]==0 && !visited[i][j]){
                    bfs(new int[] {i, j}, num++);
                }
            }
        }



        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (map[i][j] == 1) sb.append(count(i, j));
                else sb.append(0);
            }
            sb.append("\n");
        }

        System.out.println(sb.toString());


    }

    public static int count(int r, int c) {
        HashSet<Integer> set = new HashSet<>();

        int cnt = 1;
        for (int i = 0; i < 4; i++) {
            int du = r + dr[i];
            int dv = c + dc[i];

            if (du < 0 || du >= N || dv < 0 || dv >= M) continue;


            if (map[du][dv] == 0 && !set.contains(result[du][dv].num)) {
                cnt += result[du][dv].cnt;
                set.add(result[du][dv].num);
            }
        }

        return cnt % 10;
    }

    public static void bfs(int[] arr, int num) {
        Queue<int[]> queue = new LinkedList<>();
        Queue<int[]> tmp = new LinkedList<>();

        queue.add(arr);
        tmp.add(arr);

        int cnt = 1;
        visited[arr[0]][arr[1]] = true;


        while (!queue.isEmpty()) {
            int[] now = queue.poll();

            for (int i = 0; i < 4; i++) {
                int du = now[0] + dr[i];
                int dv = now[1] + dc[i];

                if (du >= 0 && du < N && dv >= 0 && dv < M && map[du][dv] == 0 && !visited[du][dv]) {
                    cnt++;
                    queue.add(new int[]{du, dv});
                    tmp.add(new int[]{du, dv});
                    visited[du][dv] = true;


                }
            }
        }

        while (!tmp.isEmpty()) {
            int[] now = tmp.poll();

            result[now[0]][now[1]].cnt = cnt;
            result[now[0]][now[1]].num = num;

        }

    }

}

File: 백준/Gold/16946. 벽 부수고 이동하기 4/벽 부수고 이동하기 4.py
from collections import deque

def bfs(i, j, num):
    q = deque([(i, j)])
    visited[i][j] = num
    cnt = 0
    while q:
        cr, cc = q.popleft()
        cnt+=1
        for di, dj in DIR:
            du = cr+di
            dv = cc+dj
            if du<0 or dv<0 or du>=N or dv>=M or visited[du][dv]:
                continue
            if arr[du][dv]==0:
                visited[du][dv] = num
                q.append((du, dv))
    cnt_dict[num] = cnt


N, M = map(int, input().split())
arr = [list(map(int, list(input()))) for _ in range(N)]
visited = [[0]*M for _ in range(N)]
cnt_dict = {}
idx = 0
DIR = (-1, 0), (0, 1), (1, 0), (0, -1)
for i in range(N):
    for j in range(M):
        if arr[i][j]==0 and not visited[i][j]:
            idx+=1
            bfs(i, j, idx)

for i in range(N):
    for j in range(M):
        if arr[i][j]==1:
            used_dict = {}
            for di, dj in DIR:
                du, dv = i+di, j+dj
                if du<0 or dv<0 or du>=N or dv>=M:
                    continue
                if visited[du][dv] !=0 and not used_dict.get(visited[du][dv], 0):
                    arr[i][j] =(arr[i][j] + cnt_dict[visited[du][dv]])%10
                    used_dict[visited[du][dv]] = 1

for k in range(N):
    print(*arr[k], sep="")